kernel/timeout.c

```diff
@@ -0,0 +1,243 @@
+
+/* During a call to z_clock_announce(), the "current" time is "ahead"
+ * of the reference used by timeout_list by this amount.
+ */
+static int announce_advance;
+
+#if defined(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME)
+int z_clock_hw_cycles_per_sec = CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
+#endif
+
+static struct _timeout *first(void)
+{
+	sys_dnode_t *t = sys_dlist_peek_head(&timeout_list);
+
+	return t == NULL ? NULL : CONTAINER_OF(t, struct _timeout, node);
+}
+
+static struct _timeout *next(struct _timeout *t)
+{
+	sys_dnode_t *n = sys_dlist_peek_next(&timeout_list, &t->node);
+
+	return n == NULL ? NULL : CONTAINER_OF(n, struct _timeout, node);
+}
+
+static void remove(struct _timeout *t)
+{
+	if (next(t) != NULL) {
+		next(t)->dticks += t->dticks;
+	}
+
+	sys_dlist_remove(&t->node);
+	t->dticks = _INACTIVE;
+}
+
+static s32_t adjust_elapsed(s32_t ticks)
+{
+	ticks -= z_clock_elapsed();
+	return ticks < 0 ? 0 : ticks;
+}
+
+void _add_timeout(struct _timeout *to, _timeout_func_t fn, s32_t ticks)
+{
+	__ASSERT(to->dticks < 0, "");
+	to->fn = fn;
+
+	LOCKED(&timeout_lock) {
+		struct _timeout *t;
+
+		to->dticks = adjust_elapsed(ticks) + announce_advance;
+		for (t = first(); t != NULL; t = next(t)) {
+			__ASSERT(t->dticks >= 0, "");
+
+			if (t->dticks > to->dticks) {
+				t->dticks -= to->dticks;
+				sys_dlist_insert_before(&timeout_list,
+							&t->node, &to->node);
+				break;
+			}
+			to->dticks -= t->dticks;
+		}
+
+		if (t == NULL) {
+			sys_dlist_append(&timeout_list, &to->node);
+		}
+	}
+
+	z_clock_set_timeout(_get_next_timeout_expiry(), false);
+}
+
+int _abort_timeout(struct _timeout *to)
+{
+	int ret = _INACTIVE;
+
+	LOCKED(&timeout_lock) {
+		if (to->dticks != _INACTIVE) {
+			remove(to);
+			ret = 0;
+		}
+	}
+
+	return ret;
+}
+
+s32_t z_timeout_remaining(struct _timeout *to)
+{
+	s32_t ticks = 0;
+
+	if (to->dticks == _INACTIVE) {
+		return 0;
+	}
+
+	LOCKED(&timeout_lock) {
+		for (struct _timeout *t = first(); t != NULL; t = next(t)) {
+			ticks += t->dticks;
+			if (to == t) {
+				break;
+			}
+		}
+	}
+
+	return ticks;
+}
+
+void z_clock_announce(s32_t ticks)
+{
+	struct _timeout *t = NULL;
+
+#ifdef CONFIG_TIMESLICING
+	z_time_slice(ticks);
+#endif
+
+	LOCKED(&timeout_lock) {
+		curr_tick += ticks;
+		announce_advance = ticks;
+	}
+
+	while (true) {
+		LOCKED(&timeout_lock) {
+			t = first();
+			if (t != NULL) {
+				if (t->dticks <= announce_advance) {
+					announce_advance -= t->dticks;
+					t->dticks = 0;
+					remove(t);
+				} else {
+					t->dticks -= announce_advance;
+					t = NULL;
+				}
+			}
+		}
+
+		if (t == NULL) {
+			break;
+		}
+
+		t->fn(t);
+	}
+
+	announce_advance = 0;
+	z_clock_set_timeout(_get_next_timeout_expiry(), false);
+}
+
+s32_t _get_next_timeout_expiry(void)
+{
+	s32_t ret = 0;
+	int max = can_wait_forever ? K_FOREVER : INT_MAX;
+
+	LOCKED(&timeout_lock) {
+		struct _timeout *to = first();
+
+		ret = to == NULL ? max : adjust_elapsed(to->dticks);
+	}
```

Comment:
Here you an still have race between `adjust_elapsed()` and the `z_clock_set_timeout()`...
We can run whole `z_clock_set_timeout(_get_next_timeout_expiry(), false);` with interrupts locked, but this will not solve potential tick "skipping" due to CPU speed (however it will make it very unlikely).

Is there a possibility to read "now()" only for newly added timeouts, and process existing ones basing only on the values from `z_clock_announce()` ?

Something like this:
- new_timeout = ms_to_ticks(timeout) + z_clock_elapsed();
- z_clock_set_timeout(new_timeout) /* reference: time of last z_clock_announce() */

Reply:
FWIW, that race is inherent: the flow of time doesn't stop.  On hardware that have a minimum-future-interrupt time, that's the responsibility of the driver to make sure that the interrupt still happens even if the kernel asks for something very soon (or already expired).  If you check the timerhacks branch, you'll see the beginnings of a rewrite of the HPET driver that has logic for this.

As for the "now()" being timeout-specific, check the logic in z_clock_announce().  I believe it does what you want.  Timeouts added during processing are done relative to the current head of the timeout queue, which is updated with each item pulled off the list.  So if you have a timeout that adds itself, and have missed a clock interrupt or three, then it should still add back a timeout relative to its scheduled expiration and not whatever the wall clock says.